File: GameBoard.cpp
#include "GameBoard.hpp"
#include <algorithm>

GameBoard::GameBoard(int size)
	: m_size(size)
{
}

bool GameBoard::PlaceShip(const Ship& ship)
{
	// Проверка на выход за границы
	for (const auto& coord : ship.GetCoordinates())
	{
		if (coord.first < 0 || coord.first >= m_size ||
			coord.second < 0 || coord.second >= m_size)
		{
			return false;
		}
	}

	// Проверка на пересечение с другими кораблями
	for (const auto& existingShip : m_ships)
	{
		for (const auto& existingCoord : existingShip.GetCoordinates())
		{
			for (const auto& newCoord : ship.GetCoordinates())
			{
				// Проверка самой координаты и соседних клеток
				for (int i = -1; i <= 1; i++)
				{
					for (int j = -1; j <= 1; j++)
					{
						if (existingCoord.first + i == newCoord.first &&
							existingCoord.second + j == newCoord.second)
						{
							return false;
						}
					}
				}
			}
		}
	}

	m_ships.push_back(ship);
	return true;
}

Ship::ShotResult GameBoard::ReceiveShot(std::pair<int, int> coord)
{
	// Проверка на повторный выстрел
	if (m_shots.find(coord) != m_shots.end())
	{
		return Ship::ShotResult::eAlreadyShot;
	}

	m_shots.insert(coord);

	// Проверка попадания
	for (auto& ship : m_ships)
	{
		if (ship.TakeHit(coord))
		{
			if (ship.IsSunk())
			{
				return Ship::ShotResult::eSunk;
			}
			return Ship::ShotResult::eHit;
		}
	}

	m_misses.push_back(coord);
	return Ship::ShotResult::eMiss;
}

bool GameBoard::IsAllShipsSunk() const
{
	for (const auto& ship : m_ships)
	{
		if (!ship.IsSunk())
		{
			return false;
		}
	}
	return true;
}

GameBoard::BoardStateType GameBoard::GetVisibleState(bool forOwner) const
{
	BoardStateType state(m_size, std::vector<char>(m_size, '.'));

	// Всегда показываем промахи
	for (const auto& miss : m_misses)
	{
		state[miss.first][miss.second] = 'O';
	}

	// Всегда показываем попадания
	for (const auto& ship : m_ships)
	{
		for (const auto& coord : ship.GetCoordinates())
		{
			if (m_shots.find(coord) != m_shots.end())
			{
				state[coord.first][coord.second] = 'X';
			}
		}
	}

	// Показываем корабли ТОЛЬКО если это поле владельца
	if (forOwner)
	{
		for (const auto& ship : m_ships)
		{
			for (const auto& coord : ship.GetCoordinates())
			{
				// Показываем неподбитые части кораблей
				if (m_shots.find(coord) == m_shots.end())
				{
					state[coord.first][coord.second] = 'S';
				}
			}
		}
	}

	return state;
}

// Кастомные размеры кораблей в зависимости от размера поля
GameBoard::ShipSizesType GameBoard::MakeShipSizes(std::array < std::pair<int, int>, 4> shipConfig)
{		
	// Формируем вектор размеров кораблей
	ShipSizesType sizes;
	for (const auto& pair : shipConfig)
	{
		int shipLength = pair.first;  
		int shipCount = pair.second;

		for (int i = 0; i < shipCount; i++)
		{
			sizes.push_back(shipLength);
		}
	}

	return sizes;
}

File: BattleshipCS/GameBoard.cs
// GameBoard.cs
namespace BattleshipCS;

public class GameBoard
{
    // Статические поля и методы
    public static readonly int DEFAULT_BOARD_SIZE = 10;
    public static readonly (int, int)[] DEFAULT_SHIP_CONFIG = {
        (4, 1), (3, 2), (2, 3), (1, 4)
    };

    // Статический метод для валидации размера доски
    public static bool IsValidBoardSize(int size)
    {
        return size >= 5 && size <= 15;
    }

    // Свойства с приватными сеттерами
    public int Size { get; }
    public List<Ship> Ships { get; private set; }
    public HashSet<(int, int)> Shots { get; private set; }
    public List<(int, int)> Misses { get; private set; }

    // Вычисляемое свойство
    public int RemainingShips => Ships.Count(ship => !ship.IsSunk());

    public GameBoard(int size)
    {
        // Валидация с выбрасыванием исключения
        if (!IsValidBoardSize(size))
            throw new ArgumentException($"Недопустимый размер доски: {size}. Допустимый диапазон: 5-15");

        try
        {
            Size = size;
            Ships = new List<Ship>();
            Shots = new HashSet<(int, int)>();
            Misses = new List<(int, int)>();
        }
        catch (Exception ex)
        {
            throw new InvalidOperationException("Ошибка инициализации игровой доски", ex);
        }
    }

    public bool PlaceShip(Ship ship)
    {
        // Защитный блок
        try
        {
            // Проверка на выход за границы
            foreach (var coord in ship.Coordinates)
            {
                if (coord.Item1 < 0 || coord.Item1 >= Size ||
                    coord.Item2 < 0 || coord.Item2 >= Size)
                {
                    return false;
                }
            }

            // Проверка на пересечение с другими кораблями
            foreach (var existingShip in Ships)
            {
                foreach (var existingCoord in existingShip.Coordinates)
                {
                    foreach (var newCoord in ship.Coordinates)
                    {
                        // Проверка самой координаты и соседних клеток
                        for (int i = -1; i <= 1; i++)
                        {
                            for (int j = -1; j <= 1; j++)
                            {
                                if (existingCoord.Item1 + i == newCoord.Item1 &&
                                    existingCoord.Item2 + j == newCoord.Item2)
                                {
                                    return false;
                                }
                            }
                        }
                    }
                }
            }

            Ships.Add(ship);
            return true;
        }
        catch (Exception ex)
        {
            throw new InvalidOperationException("Ошибка при размещении корабля", ex);
        }
    }

    public Ship.ShotResult ReceiveShot((int, int) coord)
    {
        // Защитный блок с обработкой исключений
        try
        {
            // Проверка на повторный выстрел
            if (Shots.Contains(coord))
            {
                return Ship.ShotResult.AlreadyShot;
            }

            Shots.Add(coord);

            // Проверка попадания
            foreach (var ship in Ships)
            {
                if (ship.TakeHit(coord))
                {
                    if (ship.IsSunk())
                    {
                        return Ship.ShotResult.Sunk;
                    }
                    return Ship.ShotResult.Hit;
                }
            }

            Misses.Add(coord);
            return Ship.ShotResult.Miss;
        }
        catch (Exception ex)
        {
            throw new InvalidOperationException($"Ошибка при обработке выстрела в ({coord.Item1}, {coord.Item2})", ex);
        }
    }

    public bool IsAllShipsSunk()
    {
        return Ships.All(ship => ship.IsSunk());
    }

    public char[,] GetVisibleState(bool forOwner)
    {
        var state = new char[Size, Size];

        // Инициализация поля
        for (int i = 0; i < Size; i++)
            for (int j = 0; j < Size; j++)
                state[i, j] = '.';

        // Отображение промахов
        foreach (var miss in Misses)
        {
            state[miss.Item1, miss.Item2] = 'O';
        }

        // Отображение попаданий
        foreach (var ship in Ships)
        {
            foreach (var coord in ship.Coordinates)
            {
                if (Shots.Contains(coord))
                {
                    state[coord.Item1, coord.Item2] = 'X';
                }
            }
        }

        // Показываем корабли ТОЛЬКО если это поле владельца
        if (forOwner)
        {
            foreach (var ship in Ships)
            {
                foreach (var coord in ship.Coordinates)
                {
                    // Показываем неподбитые части кораблей
                    if (!Shots.Contains(coord))
                    {
                        state[coord.Item1, coord.Item2] = 'S';
                    }
                }
            }
        }

        return state;
    }

    public static List<int> MakeShipSizes((int, int)[] shipConfig)
    {
        // Защитный блок
        try
        {
            var sizes = new List<int>();
            foreach (var (shipLength, shipCount) in shipConfig)
            {
                if (shipLength <= 0 || shipCount < 0)
                    throw new ArgumentException("Недопустимая конфигурация кораблей");

                for (int i = 0; i < shipCount; i++)
                {
                    sizes.Add(shipLength);
                }
            }
            return sizes;
        }
        catch (Exception ex)
        {
            throw new InvalidOperationException("Ошибка при создании списка размеров кораблей", ex);
        }
    }
}